dotnet/src/Svg.cs

```diff
@@ -0,0 +1,162 @@
+    /// of static initializers (because the files have to be in the same order every time). The JS uses this value as a cache breaker when
+    /// downloading any SVG icon.
+    /// </summary>
+    public static string CombinedCacheBreaker { get; private set; } = string.Empty;
+
+    /// <summary>
+    /// Gets an <see cref="SvgImage"/> for caching and reuse.
+    /// </summary>
+    /// <param name="fileName">The filename to grab, defaults to the caller's name via <see cref="CallerMemberNameAttribute"/>.</param>
+    /// <param name="isSpot">Whether the image is a Spot or not.</param>
+    /// <param name="bypassSizeCheck">Whether to allow bypassing the size check, for things we know not to repeat in a page (stil shouldn't be huge).</param>
+    /// <returns>The <see cref="SvgImage"/> to cache.</returns>
+    internal static SvgImage GetImage([CallerMemberName] string? fileName = null, bool isSpot = false, bool bypassSizeCheck = false)
+    {
+        if (string.IsNullOrEmpty(fileName))
+        {
+            return Empty;
+        }
+
+        try
+        {
+            var imageString = Helpers.GetSvg(fileName!, isSpot);
+#if DEBUG
+            // Okay so now we have the svg asset loaded into the string,
+            // if we're on DEBUG let's do some gut checking on the .svg itself.
+            // This way we throw in dev.
+            if (imageString is null)
+            {
+                throw new Exception($"Unable to find {fileName} in embedded resources.");
+            }
+
+            if (imageString.Length > MaxReasonableSize && !bypassSizeCheck)
+            {
+                throw new Exception($"{fileName} is larger than the maximum allowed SVG icon size: {MaxReasonableSize} bytes.");
+            }
+
+            if (imageString.Contains("<!--"))
+            {
+                throw new Exception($"{fileName} contains <!-- --> style comments. Please ensure this file is minified properly.");
+            }
+#endif
+            CombinedCacheBreaker = Helpers.ToSha256Hash(CombinedCacheBreaker + imageString).Substring(0, 12);
+            return new SvgImage(imageString);
+        }
+        catch (Exception ex)
+        {
+#if DEBUG
+            // If the file wasn't able to be loaded at all, we should just throw.
+            // This throw should never happen on prod since people should be checking after they add a new SVG
+            // and the entry for it in this file, but you know, just in case.
+            throw new Exception($"Unable to open {fileName}.svg (isSpot: {isSpot})", ex);
+#else
+            return Empty;
+#endif
+        }
+    }
+}
+
+/// <summary>
+/// Describes an svg image string that is sanitized for use in Html
+/// </summary>
+public class SvgImage : HtmlString
+{
+    private readonly string _original;
+    private readonly ConcurrentDictionary<(string cssClass, string title), HtmlString> cache = new();
```

Comment:
Out of curiosity: is the reason why we are introducing a cache here performance (avoid performing regex operations multiple times)?

Reply:
Yup. Partially runtime performance (cpu) and partially memory performance (string allocations). The reasoning behind this is that the icon on a page is basically _never_ going to change. So in that case, adding the cache it becomes "generate once, read many times". At the scale of Stack Overflow, the number of reads can be... significant.

This implementation was taken from the Core codebase as well.